### PR TITLE
BUG: hash_pandas_object ignores optional arguments when the input is a DataFrame.

### LINKED ISSUE #41404: BUG: hashing's are the same for different key values  for hash_pandas_object
- [x ] I have checked that this issue has not already been reported.

- [ x] I have confirmed this bug exists on the latest version of pandas.

- [ x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
# Your code here
hash_pandas_object(test[       columns_names[i]      ],    index=True, encoding='utf8', hash_key='012' ,    categorize=True)
0      3713087409444908179
1      7478705303072568462
2     12024724921319894105
3     12785939622558835299
4      9788992550609991128
5      1239052552041868816
6      9610202078597672705
7     12287384021013641209
8     10264240190786022141
9     10535148974563425818
10    10238940258630658604
11    15446383648481672096
12    14265484681526586699
13     8862960024351814462
dtype: uint64

hash_pandas_object(test[       columns_names[i]      ],    index=True, encoding='utf8', hash_key='01298768755' ,    categorize=True)
0      3713087409444908179
1      7478705303072568462
2     12024724921319894105
3     12785939622558835299
4      9788992550609991128
5      1239052552041868816
6      9610202078597672705
7     12287384021013641209
8     10264240190786022141
9     10535148974563425818
10    10238940258630658604
11    15446383648481672096
12    14265484681526586699
13     8862960024351814462
dtype: uint64


hash_pandas_object(test[     [  columns_names[i] , columns_names[j]  ]     ],    index=True, encoding='utf8', hash_key='01' ,    categorize=True)
0     11107058607426530111
1     15666232225746534312
2      1136675766145783381
3     14892489092684772659
4      8519430825150424018
5       550646855301521146
6      3846031041217881485
7      2936614219041217571
8     16182698869780262111
9      2895548739675332954
10      677258434224654732
11     6105852029672525672
12    15095703462911844621
13     6081994522921680694
dtype: uint64
hash_pandas_object(test[     [  columns_names[i] , columns_names[j]  ]     ],    index=True, encoding='utf8', hash_key='0198076674534' ,    categorize=True)
0     11107058607426530111
1     15666232225746534312
2      1136675766145783381
3     14892489092684772659
4      8519430825150424018
5       550646855301521146
6      3846031041217881485
7      2936614219041217571
8     16182698869780262111
9      2895548739675332954
10      677258434224654732
11     6105852029672525672
12    15095703462911844621
13     6081994522921680694
dtype: uint64


test[     [  columns_names[i] , columns_names[j]  ]     ]
    A  B
0   0 -1
1   0 -1
2   0  0
3   0  0
4   1  0
5   1  0
6   2  0
7   2  2
8   2  2
9   2  2
10  2  2
11  2  2
12 -1  2
13 -1  2


hashing's are the same for different key values 




```

#### Problem description

it should be different hashing values for different keys  

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here leaving a blank line after the details tag]

print("sklearn.__version__ = ",     sklearn.__version__)
sklearn.__version__ =  0.24.2

pd.show_versions()

INSTALLED VERSIONS
------------------
commit           : 2cb96529396d93b46abab7bbc73a208e708c642e
python           : 3.8.7.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19041
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 9, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United States.1252

pandas           : 1.2.4
numpy            : 1.19.5
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 20.3.3
setuptools       : 51.3.3
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.6.3
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by i-aki-y**
- [x] closes #41404
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

In this PR, I will try to solve issue #41404.
I would appreciate it if you could review this.

## About the issue

In the current `pandas.util.hash_pandas_object` function implementation, the hash_key argument is ignored when the input object type is a DataFrame.
This is a reproducible example.

```py
df = pd.DataFrame({"x": list("abc")})
new_hash_key = "a"*16
print("Series")
print(pd.util.hash_pandas_object(df["x"]).values)
print(pd.util.hash_pandas_object(df["x"], hash_key=new_hash_key).values)
print("DataFrame")
print(pd.util.hash_pandas_object(df).values)
print(pd.util.hash_pandas_object(df, hash_key=new_hash_key).values)
```

The output is here.

```
Series
[ 4578374827886788867 17338122309987883691  5473791562133574857]
[14372590553486244284 12017144478307119910    34790805896362026]
DataFrame
[ 4578374827886788867 17338122309987883691  5473791562133574857]
[ 4578374827886788867 17338122309987883691  5473791562133574857]
```

The last output should be
```
[14372590553486244284 12017144478307119910    34790805896362026]
```

## About the fix

When the input object is a DataFrame, the hash values are calculated by applying `hash_array` function for each Series.
However, optional arguments, including `hash_key`, are not passed to the hash_array function in the current implementation.
To fix this, I passed the optional arguments to the hash_array function.

### CODE REVIEW

File: pandas/tests/util/test_hashing.py [hunk 1]
```diff
@@ -255,6 +255,16 @@ def test_hash_keys():
     assert (a != b).all()
 
 
+def test_df_hash_keys():
+    # DataFrame version of the test_hash_keys
```
Comment (mzeitlin11) [inline pandas/tests/util/test_hashing.py:259]: Can you please link the issue here?

File: pandas/core/util/hashing.py [hunk 2]
```diff
@@ -139,7 +139,10 @@ def hash_pandas_object(
         ser = Series(h, index=obj.index, dtype="uint64", copy=False)
 
     elif isinstance(obj, ABCDataFrame):
-        hashes = (hash_array(series._values) for _, series in obj.items())
+        hashes = (
+            hash_array(series._values, encoding, hash_key, categorize)
```
Comment (mzeitlin11) [inline pandas/core/util/hashing.py:143]: Since `encoding` also wasn't passed before, can you add a test that ensures we respect `encoding` too?

**Review by mzeitlin11 — COMMENTED** [review]
Thanks for the pr @i-aki-y! Some quick initial comments

File: pandas/tests/util/test_hashing.py (hunk 1 above)
Comment (i-aki-y) [reply pandas/tests/util/test_hashing.py]: @mzeitlin11  Thank you for your comment. I added a link to the issue.

File: pandas/core/util/hashing.py (hunk 2 above)
Comment (i-aki-y) [reply pandas/core/util/hashing.py:143]: I added a `test_df_encoding` to confirm that hash values can change when the different encodings are specified.

**Comment by jreback** [comment]
lgtm @i-aki-y can you add a release note bug fix section (Other is fine) for 1.3 ping on green.

**Comment by i-aki-y** [comment]
@jreback Thanks, I have added a whatsnew entry.

**Comment by jreback** [comment]
thanks @i-aki-y

**Comment by jreback** [comment]
@meeseeksdev backport 1.3.x

**Pull request merged**